StyleSheet : Reserve space for `valueChanged` icon

ericmehl opened the pull request:
This fixes the `valueChanged` icon overlap with Transform tool orientation plugs. It would initially overlap the left side of the text when changing the orientation from `Parent`. If you then changed to another tool and back to the first, it would be fixed when Qt laid you the widgets again. By always reserving the space, the icon always appears correctly.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
I touched up the icon and its padding so it's a little more condensed now. I went with 2px on each side, though the icon doesn't strictly align to the pixel grid so it's more like 1.5px on each side. Not sure if that alignment is something that needs changing.

I also put in a scheme for identifying labels for output plugs and removing the indentation for those. It seems to be helping the query plugs.

Last is the color spline - I tried putting the label and value on the same line. I think it looks better, but not 100% convinced. It shrinks the available space for the color sliders a little. I haven't figured out how to get the alignment the same for the "Position" and "Value" labels - `_qtWidget().setMinimumWidth()` and `_qtWidget.setFixedWidth()` don't seem to work, but I'll keep trying on that.

Fingers crossed for CI. I initially had it targeting `main` by accident and switched it to `1.3_maintenance`, so maybe the CI didn't get rebooted properly.

It's now rebased on `main`, and targeted to `main`.

Not quite ready for a full review, I want to look at a wider variety of plugs and also sort out the alignment of the spline color values first.

johnhaddon commented on python/GafferUI/LabelPlugValueWidget.py:
```diff
@@ -65,6 +65,8 @@ def __init__( self, plug, horizontalAlignment=GafferUI.Label.HorizontalAlignment
 			formatter = self.__formatter,
 		)
 		self.__label._qtWidget().setObjectName( "gafferPlugLabel" )
+		if any( p.direction() == Gaffer.Plug.Direction.In for p in self.getPlugs() ) :
+			self.__label._qtWidget().setProperty( "gafferValueChanged", GafferUI._Variant.toVariant( False ) )
```
Can we do this in `__setValueChanged()` instead of here, so that everything is all in one place?

ericmehl replied:
I have a potential change for this in https://github.com/GafferHQ/gaffer/pull/5737/commits/61ebf0e67e906c22f4cd557de90163b8f9bc02de. It's nice to get that all in one place, but it does require a `setStyleSheet()` call in order to recognized the `padding-left` change.

It doesn't seem to have caused jumpy layouts or other issues, but I don't know if there are knock-on effects from `setStyleSheet()`.

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -78,35 +78,34 @@ def __init__( self, plug, **kw ) :
-						parenting = { "verticalAlignment" : GafferUI.VerticalAlignment.Top }
-					)
-					self.__valueField.setColorChooserVisible( True )
+			self.__positionLabel = GafferUI.LabelPlugValueWidget( plug.pointXPlug( 0 ) )
+			self.__positionField = GafferUI.NumericPlugValueWidget( plug.pointXPlug( 0 ) )
```
Because we're inside a `with column` block here, these widgets are going to first be parented to that column before being pulled out and parented elsewhere by the `addChild()` calls below. I find the structured (`with`) style of UI building much more readable than the manual `addChild()` method, because you can see the structure of the UI in the structure of the code. So I'd suggest that the lesser evil would be to use `with` when constructing the containers below, even if it means two lines that create the position widgets.

johnhaddon replied:
It's definitely possible to give a fixed width to the labels - the trick is to do what PlugWidget does, and set the width on the label inside the widget : `self.__positionLabel.label()._qtWidget().setFixedWidth( 400 )`. I think it would be worth doing this to get the green dots to align.

ericmehl replied:
Oh nice, thanks for that tip on `label()._qtWidget()`, it's helpful. I added that in https://github.com/GafferHQ/gaffer/pull/5737/commits/5cdd19a9fb0714f0d8d7db9b9a273f68e13ba497

ericmehl replied:
> I find the structured (`with`) style of UI building much more readable than the manual `addChild()` method

I agree, and definitely shouldn't be added and then removed from the parent container. I'll make that change, pending what we ultimately go with for the container choice discussed below.

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -78,35 +78,34 @@ def __init__( self, plug, **kw ) :
+				self.__valueField = GafferUI.ColorPlugValueWidget( plug.pointYPlug( 0 ) )
+				self.__valueField.setColorChooserVisible( True )
+
+			if isinstance( plug.pointYPlug( 0 ), Gaffer.FloatPlug ) :
+				row = GafferUI.ListContainer( orientation = GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 )
+
+				row.addChild( self.__positionLabel, verticalAlignment = GafferUI.VerticalAlignment.Top )
+				row.addChild( self.__positionField, verticalAlignment = GafferUI.VerticalAlignment.Top )
+
+				row.addChild( self.__valueLabel, verticalAlignment = GafferUI.VerticalAlignment.Top )
+				row.addChild( self.__valueField, verticalAlignment = GafferUI.VerticalAlignment.Top )
+			else :
+				grid = GafferUI.GridContainer( spacing = 4 )
```
Maybe you could use a single `with GridContainer()` for both cases, and then the only thing that would differ for the float and color modes would be the indices of the widgets within the grid?

ericmehl replied:
That was my initial approach, but I wasn't super happy with the layout for float splines. Here's what that looks like :
![image](https://github.com/GafferHQ/gaffer/assets/7873467/822ebe33-036f-428d-8937-db6d756a2946)

It seems odd to me that it doesn't fill the width, but maybe that's just me being used to the old style. Maybe there's a way to expand the widgets, `setSizePolicy()` sounds like a good place to try.

johnhaddon reviewed (commented):
Thanks Eric. This is looking good aesthetics-wise I think. I've made a few inline suggestions for potentially tidying up the implementation a little, and for getting the dots aligned in the RampPlugValueWidget...

ericmehl commented:
Ok, now that I know how to set the width of the label, the spline plug layouts got a lot easier. The latest push has everything squashed down and is ready for a new review.

johnhaddon commented on python/GafferUI/LabelPlugValueWidget.py:
```diff
@@ -160,16 +160,20 @@ def _hasUserValue( plug ) :
 
 		self.__label._qtWidget().setProperty( "gafferValueChanged", GafferUI._Variant.toVariant( valueChanged ) )
 		self.__label._repolish()
+		self.__label._setStyleSheet()
```
This will be the only place in Gaffer where we've needed this on anything other than a top-level widget - all other widgets just inherit from their parents. It doesn't seem to be necessary here (testing on Linux), and in fact actually seems to cause clipping of the labels in the ColorSpline UI. What problems were you seeing without it?

ericmehl replied:
Without `_setStyleSheet()` I don't get the space reserved for the green dot, so it overlaps with the label.

![image](https://github.com/GafferHQ/gaffer/assets/7873467/3beb3d54-6c3d-448a-8f0a-a4877f438bae)

I found some references online implying that padding would not be taken into account in Qt's `polish()` method. It is a little vague, but if the [Qt documentation](https://doc.qt.io/qt-5/qstyle.html#polish)'s mention of not changing geometry includes padding, then that sounds like intended behavior.

I don't seem to have any clipping of labels (on Windows) :
![image](https://github.com/GafferHQ/gaffer/assets/7873467/11c3a1e0-ace4-4c94-bfb2-66cbc645da56)

ericmehl replied:
Resolved in https://github.com/GafferHQ/gaffer/pull/5737/commits/b25b038edbb40d12dfcd42a143a8f24d26e055c5 and https://github.com/GafferHQ/gaffer/pull/5737/commits/d3aaa2f4146b4b6c848bd0ca02c368db6e1d62e4 where I added metadata to control whether or not the icon is shown, and hiding it for spline plugs.

johnhaddon commented on python/GafferUI/RampPlugValueWidget.py:
```diff
@@ -78,35 +78,43 @@ def __init__( self, plug, **kw ) :
-				if isinstance( plug.pointYPlug( 0 ), Gaffer.FloatPlug ):
-					self.__valueField = GafferUI.NumericPlugValueWidget(
-						plug.pointYPlug( 0 ),
+				with GafferUI.ListContainer( orientation = GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 ) :
```
I am finding the new layout for ColorSplines a little bit less satisfactory than the old one. I'm wondering if it would make sense to just have a way of turning off the green dots for certain widgets or plugs instead? The idea of a non-default value doesn't make much sense for spline positions/values, at least once you've started adding your own points.

Paging @murraystevenson for a second opinion.

ericmehl replied:
Just noting that we talked about this in our weekly meeting and decided to hide the icon for spline position and value plugs. Done in https://github.com/GafferHQ/gaffer/pull/5737/commits/d3aaa2f4146b4b6c848bd0ca02c368db6e1d62e4

ericmehl commented:
I think I have a good implementation of what we talked about today. For input plugs, `LabelPlugValueWidget` now recognizes metadata entry called `showValueChangedIndicator` on the plug which it is labelling. If `False`, away goes the green dot as requested.

If it's an output plug, we never set the value `gafferShowValueChangeIndicator` in the first place, so they continue to never get a green dot, or padding that messes up the label.

And lastly, the `RampPlugValueWidget` uses this new power to hide the dot on the position and value plugs.

Ready for a new look.

johnhaddon reviewed (commented):
Sorry Eric, this seems to be the PR that refuses to die! I've made two new comments inline, which I think do need addressing. In the interests of getting a release out today, I've opened https://github.com/GafferHQ/gaffer/pull/5764 with fixups for those, and a couple of other optional commits for your consideration...I'll leave it to you and @murraystevenson which bits you take...

ericmehl commented:
Thanks for the fixes in https://github.com/GafferHQ/gaffer/pull/5764, @johnhaddon. I tried to squash down and push to your branch to merge, but I don't seem to have the right permissions. I pushed back to this branch with your fixes, and dropped the speculative commit where you used a `NameLabel` instead of `LabelPlugValueWidget`.

I also rebased onto https://github.com/GafferHQ/gaffer/pull/5765 to pick up that Windows build fix.

@murraystevenson, this is ready for a final check and merge assuming all is well.

murraystevenson commented:
Thanks! Merging.
